src/tests/JIT/CodeGenBringUpTests/CastThenBinop.cs

```diff
@@ -0,0 +1,38 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+
+using System.Runtime.CompilerServices;
+
+// Test for https://github.com/dotnet/runtime/issues/13816
+public class Test
+{
+    [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.AggressiveOptimization)]
+    static int DowncastOr(int a, int b)
+    {
+        return (byte)a | (byte)b;
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.AggressiveOptimization)]
+    static long UpcastAnd(int a, int b)
+    {
+        return (long)a & (long)b;
+    }
```

Comment:
Could you also add more test cases here:
1) when ops are complex expressions with side effects
2) when a cast should throw an overflow exception

Reply:
acf28b8